### PR TITLE
Added web profiler and toolbar

### PR DESCRIPTION
**Pull request by Nyholm**
As mentioned in #3. 

This PR implements a web debug toolbar. It uses the HistoryPlugin.

### CODE REVIEW

File: Collector/MessageJournal.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    public function getTotalRequests()
+    {
+        return count($this->data['success'])+count($this->data['failure']);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function collect(Request $request, Response $response, \Exception $exception = null)
```
Comment (dbu) [inline Collector/MessageJournal.php]: is this intentionally empty? if so, please add a comment why it is there and empty.

File: DependencyInjection/HttplugExtension.php [hunk 2]
```diff
@@ -56,6 +61,10 @@ protected function configureClients(ContainerBuilder $container, array $config)
                 $first = $name;
             }
 
+            if (isset($config['_inject_journal_plugin'])) {
+                array_unshift($arguments['plugins'], 'httplug.plugin.history');
```
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: what if somebody was already adding the history plugin? should we array_search ('httpplug.plugin.history') to avoid that?

File: README.md [hunk 3]
```diff
@@ -51,14 +51,16 @@ For information how to write applications with the services provided by this bun
 | httplug.stream_factory | Service* that provides the `Http\Message\StreamFactory`
 | httplug.client.[name] | This is your Httpclient that you have configured. With the configuration below the name would be `acme_client`.
 | httplug.client | This is the first client configured or a client named `default`.
-| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services. 
+| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.history<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services.<br><br>*The History plugin is only available when the kernel runs on debug mode and `httplug.toolbar: true`. It gets automatically injected to all clients.*
```
Comment (dbu) [inline README.md]: s/build in/built in/

Comment (dbu) [inline README.md]: s/that lives/that live/  (plural)

File: DependencyInjection/HttplugExtension.php (hunk 2 above)
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: ah, read the doc now. maybe throw an InvalidConfigurationException if its explicitly specified? then again, why not allow users to activate the history plugin and do their own thing, even in prod mode if they feel its a good idea?

**Comment by dbu** [comment]
very cool! i am impressed how simple this looks!

File: DependencyInjection/HttplugExtension.php (hunk 2 above)
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: We cant do a search/replace. People should not use "our" namespace when defining services so they would create a service like `my_history_plugin`. So it is impossible to do a search because we dont know what to search for... 

If someone uses our service, then the history will be logged twice. The readme says it is added automatically. We should not bother. 

What we should care about is that what if they are not happy with our formatter? But I think that can be solved by #17.

File: Collector/MessageJournal.php (hunk 1 above)
Comment (Nyholm) [inline Collector/MessageJournal.php]: Okey, thank you. 

A upcoming PR maybe want to log which Symfony Request that made which Httplug Request. But it should be empty for now.

**Comment by Nyholm** [comment]
> very cool! i am impressed how simple this looks!

Thank you. So am I. I barley wrote any code just some clue to make `Httplug\*` classes talk with Symfony.

File: DependencyInjection/HttplugExtension.php (hunk 2 above)
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: ah, got it. maybe we should use a more specific service name for this. its not "the" history plugin, but a history plugin linked to the message journal.

File: README.md [hunk 4]
```diff
@@ -51,14 +51,16 @@ For information how to write applications with the services provided by this bun
 | httplug.stream_factory | Service* that provides the `Http\Message\StreamFactory`
 | httplug.client.[name] | This is your Httpclient that you have configured. With the configuration below the name would be `acme_client`.
 | httplug.client | This is the first client configured or a client named `default`.
-| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services. 
+| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.history<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are built in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services.<br><br>*The History plugin is only available when the kernel runs on debug mode and `httplug.toolbar: true`. It gets automatically injected to all clients.*
```
Comment (dbu) [inline README.md]: s/that lives/that live/

File: README.md [hunk 5]
```diff
@@ -51,14 +51,16 @@ For information how to write applications with the services provided by this bun
...
```
Comment (dbu) [inline README.md]: lets make an explicit note on the history plugin. that plugin needs to be set up with a journal, so you need to define your own service for it.

File: Resources/config/data-collector.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="httplug.message_journal" class="Http\HttplugBundle\Collector\MessageJournal" public="false">
+            <tag name="data_collector" template="HttplugBundle::webprofiler.html.twig" priority="200"
+                 id="httplug"/>
+        </service>
+
+        <service id="httplug.plugin.history" class="Http\Client\Plugin\HistoryPlugin" public="false">
```
Comment (dbu) [inline Resources/config/data-collector.xml]: lets use a more specific service name here to prevent confusion with generally useable plugins.

File: DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -72,6 +72,7 @@ public function getConfigTreeBuilder()
                         ->scalarNode('stream_factory')->defaultNull()->end()
                     ->end()
                 ->end()
+                ->booleanNode('toolbar')->defaultTrue()->end()
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: i would make toolbar an arrayNode with canBeEnabled and a message_formatter option to use a custom formatter. its not a plugin configuration for me. the fact that the history plugin is used is an implementation detail. if i configure my own services with history plugin instances, i need my own configuration for formatters anyways.

the other plugins can be configured and then used as they are in any client the user wants to use them.

File: DependencyInjection/HttplugExtension.php (hunk 2 above)
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: I'll make this service "internal" instead.. Im using service id `httplug.collector.history_plugin`

File: README.md (hunk 5 above)
Comment (Nyholm) [inline README.md]: I've made it "internal". Configuration for plugins should be solved by #17

File: DependencyInjection/Configuration.php (hunk 7 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: Agreed. But I would like `canBeDisabled` instead... It should be enabled by default.

Comment (dbu) [inline DependencyInjection/Configuration.php]: oh, you are right, yes. maybe add a description() saying that the toolbar is only active if the debug toolbar itself is active.

File: Collector/MessageJournal.php (hunk 1 above)
Comment (dbu) [inline Collector/MessageJournal.php]: :+1:

File: DependencyInjection/HttplugExtension.php [hunk 8]
```diff
@@ -27,6 +27,13 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('services.xml');
         $loader->load('plugins.xml');
         $loader->load('discovery.xml');
+        if ($config['toolbar']['enabled'] && $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
```
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: is this state of the art to detect whether the debug toolbar is active? what if somebody has it active in prod env or deactivated in dev env? i guess in the latter case, we just waste some cpu cycles in dev mode which is no big deal. the other might be a problem. or is debug toolbar hard coupled to kernel.debug?

File: DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -72,8 +71,14 @@ public function getConfigTreeBuilder()
                         ->scalarNode('stream_factory')->defaultNull()->end()
                     ->end()
                 ->end()
-            ->end()
-        ;
+                ->arrayNode('toolbar')
+                    ->canBeDisabled()
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->scalarNode('formatter')->defaultValue('httplug.collector.message_journal')->end()
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: erm, i think this default value is wrong. i think it should defaultNull and if set it must be a service of a class that implements Http\Message\Formatter - or am i confused?

File: DependencyInjection/HttplugExtension.php (hunk 8 above)
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: You are correct. I should redo this. 
What I really want to know is if the service `profiler` exits and if its configuration for `toolbar` is true. 

I'll investigate some more.

File: DependencyInjection/Configuration.php (hunk 9 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: No, sorry. Im confused.. You are correct.

File: DependencyInjection/HttplugExtension.php (hunk 8 above)
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: maybe look at what other bundles that plug into the toolbar do

Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: > or is debug toolbar hard coupled to kernel.debug?

No it is not. To get the toolbar you need `framework.profiler` enabled and of course the WebProfilerBundle. There is nothing that ties it to kernel.debug. So this is technically wrong, but it works for most cases. Others are doing the same.

https://github.com/egeloen/IvoryHttpAdapterBundle/blob/master/DependencyInjection/IvoryHttpAdapterExtension.php#L62

I could use a compiler pass and check for the `web_profiler.debug_toolbar.mode` parameter if it is enabled or disabled. But I think that will complicate how I inject the HistoryPlugin because that would also be done in the compiler pass. 

Question: Are we happy with postponing support for the use case where you want the debug Httplug request via the toolbar with `kernel.debug: false`?

File: DependencyInjection/Configuration.php (hunk 7 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: Thank you. I've updated the PR

File: DependencyInjection/HttplugExtension.php (hunk 8 above)
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: > maybe look at what other bundles that plug into the toolbar do

I've looked at a few. It is common that you inject a logger or a collector to a decorating class. That injection is done in a compiler pass. Some bundles (like [GeocoderBundle](https://github.com/geocoder-php/BazingaGeocoderBundle) do always inject the logger, no matter debug or environment.

Comment (sagikazarmark) [inline DependencyInjection/HttplugExtension.php]: It's always a good idea to inject logger. Fingers Crossed handler can
decide then when and what to log.

Tobias Nyholm <notifications@github.com> ezt írta (2016. január 8., péntek):

Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: > It's always a good idea to inject logger. Fingers Crossed handler candecide then when and what to log.

Yes, I meant more of a DataCollector to collect data for the profiler. That is injected all the time

File: DependencyInjection/Configuration.php [hunk 10]
```diff
@@ -34,8 +34,7 @@ public function getConfigTreeBuilder()
                     return !empty($v['classes']['client'])
                         || !empty($v['classes']['message_factory'])
                         || !empty($v['classes']['uri_factory'])
-                        || !empty($v['classes']['stream_factory'])
-                    ;
+                        || !empty($v['classes']['stream_factory']);
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: is this styleci? my idea was that adding another check only does a diff on one line, like trailing comma in arrays.

File: DependencyInjection/Configuration.php [hunk 11]
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
...
-        ;
+                ->arrayNode('toolbar')
+                    ->canBeDisabled()
+                    ->addDefaultsIfNotSet()
+                    ->info('Extend the debug profiler with inforation about requests. This could only be active when the kernel is in debug mode.')
+                    ->children()
+                        ->scalarNode('formatter')->defaultNull()->end()
+                    ->end()
+                ->end()
+            ->end();
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: same logic here, if we add another section, there will be an unnecessary diff on this line.

File: DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
                         ->scalarNode('stream_factory')->defaultNull()->end()
                     ->end()
                 ->end()
-            ->end()
-        ;
+                ->arrayNode('toolbar')
+                    ->canBeDisabled()
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: regarding kernel.debug: could we make enabled a simple field and default it to the value kernel.debug? then people could just explicitly enable/disable it if they explicitly enabled the toolbar despite kernel.debug being false. if kernel.debug is not available here, we could have a choice of `true`, `false` and `"auto"`, with auto meaning the DI extension classs looks at kernel.debug.

File: DependencyInjection/HttplugExtension.php [hunk 13]
```diff
@@ -27,6 +27,17 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('services.xml');
         $loader->load('plugins.xml');
         $loader->load('discovery.xml');
+
+        if ($config['toolbar']['enabled'] && $s = $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
```
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: so here you would do something like:

$enabled = is_boolean($config['toolbar']['enabled']) ? $config['toolbar']['enabled'] : $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug');
if ($enabled) {
...

File: DependencyInjection/Configuration.php (hunk 10 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: Yeah, this is styleci

File: DependencyInjection/Configuration.php (hunk 12 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: Sounds like a plan

File: DependencyInjection/HttplugExtension.php (hunk 13 above)
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: Thanks! I've updated the PR.

File: README.md [hunk 14]
```diff
@@ -51,15 +51,16 @@ For information how to write applications with the services provided by this bun
...
 
 \* *These services are always an alias to another service. You can specify your own service or leave the default, which is the same name with `.default` appended. The default services in turn use the service discovery mechanism to provide the best available implementation. You can specify a class for each of the default services to use instead of discovery, as long as those classes can be instantiated without arguments.*
 
+
 If you need a more custom setup, define the services in your application configuration and specify your service in the `main_alias` section. For example, to add authentication headers, you could define a service that decorates the service `httplug.client.default` with a plugin that injects the authentication headers into the request and configure `httplug.main_alias.client` to the name of your service.
 
 ```yaml
 httplug:
-  clients:
+  clients:
```
Comment (sagikazarmark) [inline README.md]: Is this a trailing space?

File: Resources/views/webprofiler.html.twig [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+{% block menu %}
+    {# This left-hand menu appears when using the full-screen profiler. #}
+    <span class="label {{ collector.totalRequests == 0 ? 'disabled' }}">
+        <span class="icon">{{ include('@WebProfiler/Icon/ajax.svg') }}</span>
+        <strong>Httplug</strong>
+    </span>
+{% endblock %}
+
+{% block panel %}
+    <h2>Httplug</h2>
```
Comment (sagikazarmark) [inline Resources/views/webprofiler.html.twig]: HTTPlug :smile:

File: Tests/Resources/Fixtures/config/full.php [hunk 16]
```diff
@@ -1,16 +1,16 @@
...
-        'uri_factory' => 'my_uri_factory',
-        'stream_factory' => 'my_stream_factory',
-    ),
-    'classes' => array(
-        'client' => 'Http\Adapter\Guzzle6\Client',
+        'uri_factory'     => 'my_uri_factory',
+        'stream_factory'  => 'my_stream_factory',
+    ],
+    'classes' => [
+        'client'          => 'Http\Adapter\Guzzle6\Client',
```
Comment (sagikazarmark) [inline Tests/Resources/Fixtures/config/full.php]: Is this styleci as well? AFAIK symfony preset unaligns double arrows, doesn't it?

Comment (dbu) [inline Tests/Resources/Fixtures/config/full.php]: Tests is ignored by styleci i think. not sure if we really want to ignore them. consistent CS in the tests sounds like a valuable thing to have.

File: DependencyInjection/Configuration.php [hunk 17]
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
                         ->scalarNode('stream_factory')->defaultNull()->end()
                     ->end()
                 ->end()
-            ->end()
-        ;
+                ->arrayNode('toolbar')
+                    ->addDefaultsIfNotSet()
+                    ->info('Extend the debug profiler with inforation about requests. This could only be active when the kernel is in debug mode.')
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: lets remove the second phrase

File: Tests/Resources/Fixtures/config/full.php (hunk 16 above)
Comment (sagikazarmark) [inline Tests/Resources/Fixtures/config/full.php]: Ah, I see. Well, the reason why I ignored tests by default was that spec tests are snake cased and function names does not contain visibility. I agree it would make sense here, so let's just remove the test paths from exclusions.

File: DependencyInjection/Configuration.php [hunk 18]
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
...
+                    ->children()
+                        ->enumNode('enabled')->values([true, false, 'auto'])->defaultValue('auto')->end()
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: and add an info() here saying that by default, the toolbar is active when kernel.debug is true, but you can force it to on or off with this option.

File: Tests/Resources/Fixtures/config/full.php (hunk 16 above)
Comment (Nyholm) [inline Tests/Resources/Fixtures/config/full.php]: Yes, That is styleci. 
Ignoring sounds like a good idea.

**Comment by Nyholm** [comment]
Thank you for your feedback. I've updated the PR.

**Comment by sagikazarmark** [comment]
Sorry, CS fixes probably introduced a few conflicts.

**Comment by Nyholm** [comment]
No worries, I've updated the PR.

**Comment by sagikazarmark** [comment]
Cool. How far is it? I am eager to try it in a project. :smile:

**Comment by Nyholm** [comment]
I'm done with the implementation, I've received a lot feedback and made some improvements. I have addressed all the issues that were brought up. Now Im waiting for @dbu to give this PR the final thumbs up.

**Comment by sagikazarmark** [comment]
Cool.

**Comment by dbu** [comment]
thanks a lot for this one, good job!

**Comment by sagikazarmark** [comment]
:+1:

**Comment by Nyholm** [comment]
Thank you for merging. I'm also very exited to try this in a project =)

File: DependencyInjection/HttplugExtension.php [hunk 19]
```diff
@@ -54,6 +66,10 @@ protected function configureClients(ContainerBuilder $container, array $config)
                 $first = $name;
             }
 
+            if (isset($config['_inject_collector_plugin'])) {
+                array_unshift($arguments['plugins'], 'httplug.collector.history_plugin');
```
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: actually, adds the plugin first. does that mean we log before anything else? that will mean that if we have the redirect plugin active, we will not see that a redirect happened. then again, if redirect was activated, we might not care about it.
but it also means we will not log the request exactly as it was sent, with base url or credentials or whatever was added by the plugins. do we maybe need a flag to decide whether to add to the beginning or to the end?

Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: If you have the CachePlugin and put the HistoryPlugin last you will not see anything... But that might be okey, since no request was actually made.

Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: true. hm. it would be cool to see whether requests where served from
cache or the actual backend. we could try to have two history plugins
and see whether a request goes to the client or not, and also see
whether there is a redirect to serve one answer.

Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: created #26 for this idea. i guess what we have now is ok, but the other thing sure would be cool.
